contrib/babelfishpg_tsql/src/procedures.c

```diff
@@ -3749,4 +3751,73 @@ remove_delimited_identifer(char *str)
 	len = strlen(str);
 	while (isspace(str[len - 1]))
 		str[--len] = 0;
-}
\ No newline at end of file
+}
+
+Datum
+sp_enum_oledb_providers_internal(PG_FUNCTION_ARGS)
+{
+	int			rc;
+	MemoryContext savedPortalCxt;
+
+	/* SPI call input */
+	StringInfoData buf;
+
+	SPIPlanPtr	plan;
+	Portal		portal;
+	DestReceiver *receiver;
+	const char* provider_name = "tds_fdw";
+
+	if(!role_is_sa(GetSessionUserId()))
+		ereport(ERROR, (errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+						errmsg("Only members of the sysadmin role can execute this stored procedure.")));
+
+	if(GetForeignDataWrapperByName(provider_name, true) == NULL){
+		PG_RETURN_VOID();
+	}
+	
+	initStringInfo(&buf);
+	appendStringInfo(&buf, "SELECT "
+							"CAST('%s' AS sys.nvarchar(255)) AS \"Provider Name\", "
+							"CAST('{' || uuid_in(md5(subquery.provider_desc::text)::cstring) || '}' AS sys.nvarchar(255)) AS \"Parse Name\", "
+							"CAST(subquery.provider_desc AS sys.nvarchar(255)) AS \"Provider Description\" "
+						"FROM ("
+							"SELECT "
+							"extversion, 'A PostgreSQL foreign data wrapper to connect to TDS databases ' || extversion AS provider_desc "
+							"FROM pg_catalog.pg_extension "
+							"WHERE extname = '%s'"
+						") subquery;"
+			, str_toupper(provider_name, strlen(provider_name), C_COLLATION_OID), provider_name);
+
+	savedPortalCxt = PortalContext;
+
+	if (PortalContext == NULL)
+		PortalContext = MessageContext;
+
+	if ((rc = SPI_connect()) != SPI_OK_CONNECT)
+		elog(ERROR, "SPI_connect failed: %s", SPI_result_code_string(rc));
+	PortalContext = savedPortalCxt;
+
+	if ((plan = SPI_prepare(buf.data, 0, NULL)) == NULL)
+		elog(ERROR, "SPI_prepare(\"%s\") failed", buf.data);
+
+	if ((portal = SPI_cursor_open(NULL, plan, NULL, NULL, true)) == NULL)
+		elog(ERROR, "SPI_cursor_open(\"%s\") failed", buf.data);
+
+	pfree(buf.data);
+
+	receiver = CreateDestReceiver(DestRemote);
+	SetRemoteDestReceiverParams(receiver, portal);
+
+	/* fetch the result and return the result-set */
+	PortalRun(portal, FETCH_ALL, true, true, receiver, receiver, NULL);
+
+	receiver->rDestroy(receiver);
+
+	SPI_cursor_close(portal);
+
+	if ((rc = SPI_finish()) != SPI_OK_FINISH)
```

Comment:
do we not need any error handling here? will we handle errors gracefully in any of the above steps?

Reply:
Surrounded with TRY...CATCH block